### ISSUE TITLE
Publish LOOT as a Flatpak package

### ISSUE THREAD

**Issue by RacerBG** [loot/loot#1757]
Description:
Adding Flatpak support and submitting it to the Flathub website.

Additional context:
https://flathub.org/
https://docs.flatpak.org/
https://github.com/flathub/flathub/wiki/App-Submission
https://github.com/flathub/flathub/wiki/App-Requirements

**Comment by Ortham** [comment]
This would be nice to have, but it's not something I personally plan to pick up. If anyone does want to do this, I'd be interested in seeing a pull request for the necessary changes.

**Comment by Ortham** [comment]
I've started to investigate what publishing LOOT as a Flatpak package would involve, and made some notes.

## Runtimes

The KDE runtime is probably most suitable as it includes the Qt framework, so LOOT wouldn't need to bundle that.

## Building

It looks like flatpak-builder can run CMake, but at least to start with it's probably better to build LOOT and then build the flatpak using the built binaries as input: this might be a problem if the flatpak runtime libraries differ from the libraries used to build LOOT, but it looks like Qt can be matched up (there's a KDE runtime based on Qt 6.5).

I'm not sure whether the KDE runtime includes the ICU libraries or Intel TBB. If they do then we could also skip bundling them so long as we can build against the correct versions.

## Sandboxing

Flatpak runs applications in sandboxes, which means LOOT needs to ask for permissions to use system resources, and may need to make some code changes.

### Filesystem

Games could be stored anywhere, so LOOT needs general filesystem access, which can be given using the `--filesystem=host` sandbox permission, though that doesn't grant access to many system directories.

The following functions use specific paths that are not covered by the `host` permission:

- `OpenInDefaultApplication()`
    - tries to run `/usr/bin/xdg-open`
    - could replace with `QDesktopServices::openUrl()`
- `GetDriveRootPaths()`
    - calls `setmntent` with `/proc/self/mounts`
    - could maybe replace with `QStorageInfo::mountedVolumes()`, though I'm not sure that would work
- `CopyToClipboard()`
    - calls `echo` (`/usr/bin/echo`) and pipes the output to `xclip` (`/usr/bin/xclip`)
    - could replace with `QClipboard::setText()`
- `getExecutableDirectory()`
    - calls `readlink` on `/proc/self/exe`
    - tricky, this is used to get the fallback path for `getLocalAppDataPath()`, but also to get the LOOT docs and resources paths. Qt has `QCoreApplication::applicationDirPath()`, but that tries to get the path from `/proc` and falls back to assuming `argv[0]` contains the absolute filename of the executable, and I'm not sure if that's true or useful in a Flatpak.

Those functions may need to be changed, but allowing access to those specific paths might also work, e.g.

```
--filesystem=host
--filesystem=/proc/self/exe
--filesystem=/proc/self/mounts
--filesystem=/usr/bin/xdg-open
--filesystem=/usr/bin/echo
--filesystem=/usr/bin/xclip
```

Info about permissions [here](
https://docs.flatpak.org/en/latest/sandbox-permissions.html#filesystem-access).

If making code changes, LOOT could use the [XDG desktop portal API](https://flatpak.github.io/xdg-desktop-portal/), but it's probably better just to use Qt.

### Environment Variables

The following functions use environment variables:

- `getLocalAppDataPath()`
    - calls `getenv` with `XDG_CONFIG_HOME` and `HOME`
    - may be able to replace with `QStandardPaths::GenericConfigLocation`
    - LOOT mostly uses this for LOOT's own app data, but it does show up as part of Epic and Microsoft Store game detection, though I don't think they matter on Linux: you can't use the Microsoft Store on Linux and while I think you can install EGS games, the path is for a game's local app data path, which would be under Wine, not native Linux.
- `getUserProfilePath()`
    - calls `getenv` with `HOME`
    - could replace with `QStandardPaths::HomeLocation` or `QDir::homePath()`. However, `$HOME` is not replaced by flatpak.

Flatpak overrides `XDG_CONFIG_HOME` to be `~/.var/app/<app-id>/config` instead of `$HOME/.config`. It apparently doesn't change `HOME`.

## extra-data

https://docs.flatpak.org/en/latest/conventions.html#exporting-through-extra-data

I thought that we could use this as an equivalent to the installer's ability to download masterlists, but it requires the SHA-256 hashes of the files to be specified, and we don't know what they are at build time.

**Comment by Ortham** [comment]
I've managed to get CI to build a single-file flatpak bundle in the `linux-artifact` branch, and gotten it to run in a Manjaro VM against a mock Morrowind install. I've noticed a few issues:

- Opening the documentation by calling `xdg-open` opens the browser to the correct path but it doesn't have permission to view the docs, which is probably because they're inside the flatpak sandbox. Strangely, when I instead use `QDesktopServices::openUrl`, it tries to open the path of the docs inside the sandbox (i.e. starting with `/app`). I'm not sure what the correct approach for this is.
- When window loses focus the general info card text doesn't change colour apart from messages - it and all other text greys out. The same is true for plugin cards (can't remember if their messages change). This does seem to be Flatpak-specific, as it doesn't happen when the native Linux build is ran without being packaged.
- Selecting the dark theme applies the bundled dark theme, which looks even more different and worse than the default theme than on Windows. Qt's apparently got pretty good support for using the OS dark theme on Linux desktop, so we should use that instead of the bundled theme.
- Updating all master lists with no games installed crashes, the code might assume that there is at least one game configured. It works with one game configured. I don't think this is Linux or Flatpak related, I suspect I just haven't accounted for the possibility of only having one task in the `ParallelTaskExecutor`.
- The LOOT icon isn't displayed in the About dialog. This is a general Linux build issue, not Flatpak-specific, though the icon to load will be stored in a different place in the Flatpak package vs in the archive.

Also, the font used for UI text is different in the Flatpak application vs. without being packaged. This isn't an issue, as the Flatpak's font matches the general OS UI font, but I thought is worth noting.

Other Flatpak-specific areas with room for improvement are:

- Build libloot and loot against ICU v71.1 so ICU doesn't need to be bundled
- Move docs and translations so that they aren't in `/app/bin/`. They should probably be in `/app/share/io.github.loot.loot/`, though that would only apply to Flatpak packages, not Linux archives.
- The release job should probably publish the flatpak to Flathub, though that needs some setup on Flathub beforehand.
- The single-file export provided as a CI artifact seems easier to use than a sideload repo, but it does require the KDE runtime to be installed manually beforehand, so that should be documented in LOOT's readme.

**Comment by Ortham** [comment]
## Theming

With the Adwaita theme installed through Flatpak and the OS using its light theme, LOOT's default theme looks OK:

![image](https://github.com/loot/loot/assets/2430738/a22693ad-4f20-45b7-b04b-18f1b0644794)

However, LOOT's dark theme is effectively broken:

![image](https://github.com/loot/loot/assets/2430738/1bf0e657-92c9-4fb7-9203-712b1bf08794)

Switching back to LOOT's default theme and changing the OS to its dark theme looks much better, though some text colours need to be adjusted:

![image](https://github.com/loot/loot/assets/2430738/a7900c05-fec0-46d6-976f-6a9b43b7c232)

So I think LOOT should just use the OS theme plus its default styling, plus a few adjustments if the OS theme is dark. It should be possible to detect if the OS theme is light or dark using `QStyleHints::colorScheme`. 

The OS theme can change while LOOT is running, so LOOT will need to detect any change so it can adapt its styling as necessary. That can be done using `QStyleHints::colorSchemeChanged`. 

Ideally, LOOT would offer three theme choices:

- OS current theme, plus LOOT's default styling, adapted as necessary
- OS dark theme, plus LOOT's adapted default styling
- OS light theme, plus LOOT's default styling

However, I don't know how feasible that is.

**Comment by Ortham** [comment]
I've implemented checking `QStyleHints::colorScheme` and while it's reliable at detecting the OS scheme on my Manjaro KDE VM, it's much less so on my Ubuntu VM. I don't know if that's because the latter is Gnome-based or maybe just because it's older and maybe has bugs that have since been fixed.

However, theme changes while LOOT is running are not detected. I'm not sure why, I can see the `org.kde.kdeglobals.General` `ColorScheme` event being fired in dbus and that's what Qt listens for. It might be a flatpak sandboxing thing. I'm going to file that as a separate enhancement as I don't think it's significant enough to block this.
